server/src/main/java/org/elasticsearch/script/ScriptContextInfo.java

```diff
@@ -0,0 +1,380 @@
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.script;
+
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+
+public class ScriptContextInfo implements ToXContentObject, Writeable {
+    public final String name;
+    public final ScriptMethodInfo execute;
+    public final Set<ScriptMethodInfo> getters;
+
+    private static String NAME_FIELD = "name";
+    private static String METHODS_FIELD = "methods";
+
+    // ScriptService constructor
+    ScriptContextInfo(String name, Class<?> clazz) {
+        this.name = name;
+        this.execute = ScriptMethodInfo.executeFromContext(clazz);
+        this.getters = Collections.unmodifiableSet(ScriptMethodInfo.gettersFromContext(clazz));
+    }
+
+    // Deserialization constructor
+    ScriptContextInfo(String name, List<ScriptMethodInfo> methods) {
+        this.name = Objects.requireNonNull(name);
+        Objects.requireNonNull(methods);
+
+        String executeName = "execute";
+        String getName = "get";
+        // ignored instead of error, so future implementations can add methods.  Same as ScriptContextInfo(String, Class).
+        String otherName = "other";
+        Map<String, List<ScriptMethodInfo>> methodTypes = methods.stream().collect(Collectors.groupingBy(
+            m -> {
+                if (m.name.equals(executeName)) {
+                    return executeName;
+                } else if (m.name.startsWith(getName) && m.parameters.size() == 0) {
+                    return getName;
+                }
+                return otherName;
+            }
+        ));
+
+        if (!methodTypes.containsKey(executeName)) {
```

Comment:
We use `== false` instead of `!` because the former is easier to quickly see when looking at code.

Reply:
Done, even though it makes IntelliJ sad.